proto/spacemesh/v1/app_types.proto

```diff
@@ -0,0 +1,129 @@
+}
+
+message ValuesArray {
+    repeated bytes values = 1;
+}
+
+// The meta-data and value of an app's variable
+message Variable {
+    VariableMetadata meta_data = 1;
+    oneof data {
+        bytes value = 2; // since bytes size is arbitrary - this will support future var types
+        ValuesArray values = 3;
+    }
+}
+
+// The address of an app template in the templates global state store
+message TemplateAddress {
+    bytes address = 1;
+}
+
+// Information about an app template in the global state store
+message Template {
+    TemplateAddress address = 1; // template address
+    string name = 2; // template name
+
+    // note - the author can be obtained from the deploy transaction but it is nice to have it here
+    AccountId author = 3; // template uploader
+
+    // The layer can be obtained from the tx receipt but it is convenient to have it here
+    LayerNumber deploy_layer = 4; // layer that the template was deployed - 0 for genesis templates
+    TransactionId deploy_tx_id = 5; // the id of the deploy template transaction. Empty for genesis templates.
+    bytes code = 6; // template wasm code
+    repeated VariableMetadata vars = 7; // template variables order and metadata
+}
+
+// A Spacemesh app
+// Note that both layer and spawner can be obtained from the transaction and transaction id for apps spawned by a transaction
+// but is convenient to return them directly as part of apps' metadata.
+message App {
+    AccountId address = 1; // App's address in global space
+    string name = 2; // short descriptive name
+    TemplateAddress template_address = 3; // App's template address
+    LayerNumber spawn_layer = 4; // Layer in which app was spawned
+    TransactionId spawn_tx_id = 5; // Transaction which spawned the app. empty for genesis apps.
+    AccountId spawner = 6; // App spawner account
+    repeated Variable variables = 8; // ordered array of app vars and their values
+}
+
+// A request to return a template by address from the template global state store
+message GetTemplateRequest {
+    TemplateAddress address = 1; // template address
+    bool include_code = 2; // when true, template result will include full wasm code
+}
+
+// A response with AppTemplate data
+message GetTemplateResponse {
+    Template template = 1;
+}
+
+// A request to return all templates from the templates global state store with pagination
+message GetTemplatesRequest {
+    uint32 max_results = 1; // max number of results to return
+    uint32 offset = 2; // query offset
+    bool include_code = 3; // when try templates results will include full wasm code
+}
+
+// A response includes total templates available in store and a list of zero ore more templates
+message GetTemplatesResponse {
+    uint32 total_results = 1; // total number of templates
+    repeated Template templates = 2;
+}
+
+// A request to get an app's variable by var id
```

Comment:
Instead of "variable" can we give these a more descriptive name, e.g., state entry?

Reply:
@YaronWittenstein said these are called "storage variables" in SVM